### DISCUSSION TITLE
Make it easy to share / install third-party scripts: bb install

### DISCUSSION THREAD

**Discussion by borkdude** [babashka/babashka#1307] [Ideas]
Deno has an interesting feature where you can install remote scripts as a command line tool:

```
$ deno install --allow-net --allow-read --name file-server https://deno.land/std@0.146.0/http/file_server.ts
```

This creates a shim script in:

```
/Users/borkdude/.deno/bin/file-server
```

with the following contents:

```
#!/bin/sh
# generated by deno install
exec deno run --allow-read --allow-net 'https://deno.land/std@0.146.0/http/file_server.ts' "$@"
```

On Windows, in `.deno\bin\file-server.cmd`:

```
% generated by deno install %
@deno "run" "--allow-read" "--allow-net" "https://deno.land/std@0.146.0/http/file_server.ts" %*
```

We could have a similar story for babashka where you can install `.clj` scripts from a remote repository.

For single standalone scripts this would be quite easy:

```
$ bb install --name new https://gist.githubusercontent.com/borkdude/2ee8e49803b40cf0fe284c9921388f87/raw/30c01517f74ff350a89a5ddcb058a713971608b0/new
```

but as soon as those scripts have dependencies, it might become more complex: here installing an uberjar created with `bb uberjar` is probably the best solution:

```
$ bb install --name http-server https://github.com/babashka/http-server/releases/download/v0.1.11/http-server.jar
```

Another option would be to let bb clone the entire repo as a tool:

```
$ bb install --name http-server --tag v0.1.11 https://github.com/babashka/http-server
```

with some entrypoint `main.clj` or so - or you would just be able to leverage tasks from `bb.edn` from that repo to run directly:

```
$ bb http-server serve --port 1338
```

```
bb install --tag v0.1.0 --name build https://github.com/weavejester/build
build lint ...
build jar ...
```

So: `.clj` is assumed to be a single standalone script, `.jar` is assumed to be a bundled dependency, other extensions are assumed to be a repo to be cloned and invoked with tasks?

**Comment by jcooley** [comment]
Nice idea. It's interesting how it passes sandbox escape permissions as part of the install. This could be done for the JVM but I can't think of an application of it outside of applets. It might be good to version it in case you decide to break compatibility with flags passed and disk layout after more people try it out.

**Reply by borkdude** [reply]
The sandbox escape only applies to deno, bb has no such thing.

**Comment by teodorlu** [comment]
Interesting.

Really liking the idea of `bb` managing "scripts on path". I would also be interested using this to manage local local scripts. I like keeping Clojure files with `.clj` extensions, but for CLI usage, the `.clj` bothers me.

Example if `bb install` should determine if its argument is an URI or a file path itself (like slurp):

    bb install --name play ~/dev/teodorlu/play.teod.eu/play.clj

No comments on dependencies -- I haven't used those yet. All dependencies I'm using for now are in the script with `babashka.deps/add-deps`.

Personally I prefer `~/.local/share/babashka/bin` to `~/.babashka/bin`, but that might just be me. My home dir is cluttered enough as it is.

**Reply by borkdude** [reply]
I think we should respect XDG_ stuff.

**Comment by teodorlu** [comment]
It would be nice if there was a way for the remote to provide information about available versions, so that you could ask "are there updates?", "I'd like to update this thing" or "I'd like to update everything" locally.

**Reply by rads** [reply]
On the Clojure CLI side, there are a number of [third-party libraries that can be used for this](https://github.com/clojure/tools.deps.alpha/wiki/Tools#deps-management). I think it makes sense to stick to this decoupled approach for Babashka. Let's say we update [antq](https://github.com/liquidz/antq) to work with `bb install`:
```
$ bb install io.github.liquidz/antq
$ antq outdated --check-babashka-scripts --upgrade
```
The Clojure CLI tools use `~/.clojure/tools/foo.edn` to store version info, with one EDN file per tool, and [this is what `antq` uses to implement `--check-clojure-tools`](https://github.com/liquidz/antq#supported-files). Maybe we could follow a similar format like `~/.babashka/install/foo.edn`?

**Comment by teodorlu** [comment]
I would also like there to be a file I can track in my dotfiles with all my scripts.

1. Updating is changing the file
2. There's an `install` or `sync` operation that ensures you have all the available scripts of the provided versions in the file available.

Could for example be `~/.config/babashka/scripts.edn`.

**Reply by rads** [reply]
I think we should try to reuse existing patterns to manage dependencies with `deps.edn`/`bb.edn` as opposed to introducing a new file such as `scripts.edn`. For project-local scripts, you can add deps to your `bb.edn` and [call functions with `exec`](https://blog.michielborkent.nl/babashka-tasks-meets-babashka-cli.html):
```clojure
{:deps {org.babashka/http-server {:mvn/version "0.1.11"}}
 :tasks {:requires ([babashka.http-server])
         http-server (exec 'babashka.http-server/-main)}}
```
Then you can do `bb http-server` in your project, similar to `npx http-server` for `npm`. This already works today.

For global scripts, maybe running these commands in an installer script is enough? Here is an example using my shortened `bb install` suggestion:
```
$ cat ./install-bb-scripts
#!/usr/bin/env bash
bb install org.babashka/http-server
bb install io.github.borkdude/jet --tag 0.2.20
bb install io.github.rads/file-watcher --latest-sha

$ ./install-bb-scripts
org.babashka/http-server is already installed but outdated (so it will be upgraded).
io.github.borkdude/jet is already installed
io.github.rads/file-watcher is already installed but outdated (so it will be upgraded).
```

**Reply by rads** [reply]
Another option would be to enable third-party libraries like `antq` to see `bb install` deps and update them (see my comment above: https://github.com/babashka/babashka/discussions/1307#discussioncomment-3388024).

**Comment by zerg000000** [comment]
That's very similar to clojure cli. maybe we already can do it with `deps.clj`?

```
clojure -Ttools install io.github.polyfy/polylith '{:git/tag "v0.2.14-alpha" :deps/root "projects/poly"}' :as poly
```

o... one difference would be the ability to sandbox the script

**Reply by rads** [reply]
Just repeating here for clarity: the sandboxing is a deno-only feature and doesn't apply to Babashka.
https://github.com/babashka/babashka/discussions/1307#discussioncomment-3068589

**Reply by rads** [reply]
On the subject of "maybe we already can do it with `deps.clj`", I think it would be useful to try to have similar paths and data structures on the filesystem.

For example, we already have `bb.edn` which is analogous to `deps.edn` and supports the same `:deps` syntax. For this case we could have something similar to `~/.clojure/tools/foo.edn` files like `~/.babashka/install/foo.edn`.

Beyond sharing data formats, I'm not sure it would make sense to use existing code from `deps.clj` here. Also, I think the `clojure -Ttools` interface is not great for humans and not worth copying directly since `bb install` is a different use case. That's just my opinion of course.

**Comment by rads** [comment]
Would it be possible to support inferring the URL from the lib name, with similar options as `neil new` for overriding this behavior?
```
bb install io.github.babashka/neil --latest-sha
neil new --name clojure-app
```
In this case we would still be installing a top-level script even though it's using a GitHub URL internally. For completeness we'd also want to support all the ["lib to url" cases](https://github.com/babashka/neil/issues/79) here.

**Reply by borkdude** [reply]
Not sure about this: where is the name of the script for example in 

```
bb install io.github.babashka/neil --latest-sha
```

?

**Reply by teodorlu** [reply]
Perhaps add an install directive to `bb.edn`?

Not sure whether we can assume "there's one script per repo". Though we could support mulitple scripts with a way to find the different scripts. Example:

```clojure
:install {:neil "./neil"
          :neil-helper-thing "./neil_helper_thing.clj"}

;; installs `neil` by copying the `neil` file, and `neil-helper-thing` copying `neil_helper_thing.clj`
```

Enumerating existing babashka scripts that could be "installable" might be a good idea. See what makes sense.

**Reply by rads** [reply]
@borkdude: Something like @teodorlu proposed above in the `bb.edn` file seems like a reasonable solution to your question of "where is the name of the script". Let the package owners decide a default and allow users to override if needed. This is essentially the [same thing as the `bin` field in `package.json`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#bin), which is [how the `shadow-cljs` CLI tool is currently implemented](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#bin) via `npm`.

Consider the current install instructions for `shadow-cljs`, which is one of the most popular CLI tools in the Clojure community:
```
npm install -g shadow-cljs
```
(I'm showing `-g` instead of `--save-dev` here because the project-local installation is already supported by a Babashka task like `bb shadow-cljs` in `bb.edn`. This example is just about top-level scripts installed to the `PATH`.)

Let's say we wanted to migrate the CLI package to Babashka instead of CLJS. With the existing proposal, the `bb install` command feels like a downgrade in usability:
```
bb install --name shadow-cljs --tag v2.19.8 https://github.com/thheller/shadow-cljs
```
I think updating the README to use this command would be a hard sell for existing users of `shadow-cljs`. The `npm` version is already simpler, so why bother? It's the same reason I'm putting in the [effort to make `neil new` work with Kit](https://github.com/kit-clj/kit/issues/52): _the user experience of `neil new` is much better than `clj -Tnew` due to the simpler interface_. [1]

If the command looks like this, the usability is on-par with NPM (which is good) [2]:

```
bb install io.github.thheller/shadow-cljs
```
Is it harder to implement this? Yes, but I think it's worth it over having to hunt down the GitHub URL and tag every time we want to remember how to install anything. Since improving usability here is about improving defaults, we could still support overriding these defaults as user-provided options.

[1] Even though we were able to improve the `clj -Tnew` API with a wrapper script (`neil`), this approach doesn't really work for `neil` itself, which is what `bb install` is supposed to help with.

[2] We do have to add the fully-qualified package (`io.github.thheller/shadow-cljs`) but I think this is a reasonable tradeoff given that Clojure CLI tools don't allow unqualified symbols for lib names. It also means we don't need to rely on a global registry like NPM does.

**Reply by borkdude** [reply]
I think you're missing one thing here: bb install is for babashka scripts/projects, not a general replacement for npm or brew.

**Reply by rads** [reply]
@borkdude: Got it. I can empathize with that since I know we want to keep things simple. For me personally, I will still feel the need to package all my future `bb` scripts with `brew` and various package managers to get a better installation experience for end users, and I was hoping `bb install` would make it so I wouldn't have to do that.

I think [the comment in this Reddit thread](https://www.reddit.com/r/Clojure/comments/w1hfm3/comment/igmcf3i/) applies just as much to `bb install` here as it does `clojure -Ttools`. Babashka is not the same tool as NPM or Homebrew but it's the same use-case here: we want to install a program that we can access system-wide with as little friction as possible.

Even for `neil`, it would seem like a regression to go from an installation command of `brew install babashka/brew/neil` to `bb install --name neil --tag v0.1.43 https://github.com/babashka/neil`. It would be less work for us as implementors at the cost of making the end user experience more complicated.

If I'm totally off the mark, and `bb install` will never solve this problem, I think there are other ways to make distribution easier for `bb` scripts without `bb install`. For example, I could make templates for `brew`/`scoop`/`nix` for `bb` projects so an "easy" installation command is not so hard to implement in the future for `bb` script maintainers. That said, I wanted to get aligned on your end goals for `bb install` to make sure I'm not wasting any time on that effort.

**Reply by borkdude** [reply]
I'm not saying we shouldn't have a good UX: I think you're right about that, but the primary goal of bb install is an easy way to install bb script, not a generic package manager.

**Reply by rads** [reply]
Sounds good 🙂 I think I've written a lot already so I'll sit back and follow the thread to see where we end up here. Thanks for your feedback as always!

**Reply by rads** [reply]
One middle-ground option I just thought of: I can make a `brew` package that runs `bb install` with a specific version and the full GitHub URL. That way the installation commands remain simple (e.g. `brew install rads/brew/file-watcher`) but I could still benefit from not having to use something like `bb gen-script` to produce artifacts for package managers.

**Comment by teodorlu** [comment]
Here's a draft that works for local scripts. Local scripts must have a `bb.edn` file, and scripts must be runnable with `bb -m MAIN` from that folder.

Works by generating a shell script wrapper in `~/.local/bin` and making that wrapper executable.

Not sure whether this contributes anything to the discussion, but there's working code and a CLI draft.

https://play.teod.eu/bb-install/

**Comment by rads** [comment]
@borkdude @teodorlu: This last week I've been working on a "batteries-included" tool to install all existing Babashka projects and scripts, with support for both local and remote deps. I took care to ensure there is no external metadata in my implementation, meaning all the state is in the `~/.babashka/bin` folder and each script is self-contained with its own EDN metadata for `bbin ls`. This is nice because you can copy-paste a file in `~/.babashka/bin` to another machine without problems, and then `bbin` can update that script on the new machine without any external registry.

I've been using this tool myself to install all the Babashka scripts and projects I use. I was able to reuse the deps inference functions from `neil new` to make it easier to install things. It's pretty exciting to be able to type `bbin install io.github.rads/watch` and then `watch -h` _just works_ right after that. The key thing here is that I don't have to invest anything as a project maintainer to take advantage of this, yet I also don't need to go out of my way as a project _consumer_ to accommodate this. For me personally, this will allow me to release smaller, more focused libraries since it removes the burden of packaging all these things separately.

I think even if this `bbin` command is too high-level to be included in Babashka, I will likely keep maintaining it going forward because it's such a useful tool even without being part of the Babashka core. **My plan is to finish up the polish and docs this weekend so I can do a `0.0.1` release. Then you folks can try this out for yourselves!**

Here's an example of how it looks. These commands are fully functional for me with my unreleased code right now:
```
$ brew install rads/brew/bbin

$ bbin help
Usage: bbin <command>

  bbin install    Install a script
  bbin uninstall  Remove a script
  bbin ls         List installed scripts
  bbin bin        Display bin folder

$ bbin bin
/Users/rads/.babashka/bin

$ bbin install io.github.babashka/neil --main-opts '["-f" "neil"]'
{:coords {:git/sha "fb9b9a0a432089b5c93868b45e898646ce74504e",
          :git/tag "v0.1.43",
          :git/url "https://github.com/babashka/neil"},
 :lib io.github.babashka/neil}

$ bbin install io.github.babashka/neil --local/root ~/src/neil --main-opts '["-f" "neil"]' --as neil-dev
{:coords {:local/root "/Users/rads/src/neil"}, :lib io.github.babashka/neil}

$ bbin install https://gist.githubusercontent.com/rads/da8ecbce63fe305f3520637810ff9506/raw/e83305656f2d145430085d5414e2c3bff776b6e8/portal.clj
{:coords {:http/url "https://gist.githubusercontent.com/rads/da8ecbce63fe305f3520637810ff9506/raw/e83305656f2d145430085d5414e2c3bff776b6e8/portal.clj"}}

# Open a Portal window with all installed scripts
$ portal <(echo `bbin ls --no-color`)
```
```clojure
;; Output of `bbin ls` with my current installed scripts:

{http-server {:coords {:git/sha "3906760b88d8aa77237f6537c44dd84669179edd",
                       :git/tag "v0.1.11",
                       :git/url "https://github.com/babashka/http-server"},
              :lib io.github.babashka/http-server},
 neil {:coords {:git/sha "fb9b9a0a432089b5c93868b45e898646ce74504e",
                :git/tag "v0.1.43",
                :git/url "https://github.com/babashka/neil"},
       :lib io.github.babashka/neil},
 neil-dev {:coords {:local/root "/Users/rads/src/neil"},
           :lib io.github.babashka/neil},
 portal {:coords {:http/url "https://gist.githubusercontent.com/rads/da8ecbce63fe305f3520637810ff9506/raw/e83305656f2d145430085d5414e2c3bff776b6e8/portal.clj"}},
 watch {:coords {:git/sha "ad6076d3b268efd81c555a246ff0b2fd36350ca6",
                 :git/tag "v0.0.3",
                 :git/url "https://github.com/rads/watch"},
        :lib io.github.rads/watch}}
```

(I'm still deciding if I should call it `bpm` instead of `bbin`. 😄)

**Reply by rads** [reply]
Here's an example of what the auto-generated `~/.babashka/bin/watch` file looks like after running `bbin install io.github.rads/watch` with no other options:
```bash
#!/usr/bin/env bash
set -e

# :bbin/start
#
# {:lib io.github.rads/watch,
#  :coords
#  {:git/url "https://github.com/rads/watch",
#   :git/tag "v0.0.3",
#   :git/sha "ad6076d3b268efd81c555a246ff0b2fd36350ca6"}}
#
# :bbin/end

PACKAGE_ROOT='/Users/rads/.gitlibs/libs/io.github.rads/watch/ad6076d3b268efd81c555a246ff0b2fd36350ca6'
PACKAGE_LIB='io.github.rads/watch'
PACKAGE_COORDS='{:git/url "https://github.com/rads/watch", :git/tag "v0.0.3", :git/sha "ad6076d3b268efd81c555a246ff0b2fd36350ca6"}'
PACKAGE_MAIN_OPTS_FIRST='-m'
PACKAGE_MAIN_OPTS_SECOND='rads.watch'

if [[ ! -d "$PACKAGE_ROOT" ]]; then
  bb --config <(echo "{:deps {$PACKAGE_LIB $PACKAGE_COORDS}}") -e
fi

exec bb \
  --deps-root "$PACKAGE_ROOT" \
  --config "$PACKAGE_ROOT/bb.edn" \
  $PACKAGE_MAIN_OPTS_FIRST "$PACKAGE_MAIN_OPTS_SECOND" \
  -- "$@"
```
You can install this locally without `bbin` by copy-pasting the code into an executable file in your `PATH`.

**Reply by borkdude** [reply]
@rads Looks good and I'm looking forward to trying this out. I'd appreciate if your tool didn't use `~/.babashka` but `~/.bbin` as `.babashka` is reserved by babashka itself. I think I like the `bbin` name better. Why do you think this is "too high level" for bb? And what does `--main-opts` do?

**Reply by rads** [reply]
@borkdude: Yeah no problem, I can easily change the path to `~/.bbin`.

I'd also like to include some metadata for `bbin` inside a project's `bb.edn` file (see the `--main-opts` explanation below). Right now [I'm putting that in a top-level `:bin` key](https://github.com/rads/watch/blob/ad6076d3b268efd81c555a246ff0b2fd36350ca6/bb.edn#L8), but I should probably rename that since `bbin` is a third-party tool. What do you think would make sense for some tool-specific metadata in the `bb.edn` file? Maybe `:bbin`? Or would a namespaced key be better like `:bbin/bin`?

> Why do you think this is "too high level" for bb?

Just a feeling I got from the [earlier thread](https://github.com/babashka/babashka/discussions/1307#discussioncomment-3353164) here when proposing the simplified command. After that discussion, I felt like the best way to show my intent was to build something that's easy to use and has defaults that make sense. My code will be MIT and you are welcome to take whatever you want for Babashka (or even better, the whole thing!).

> And what does --main-opts do?

In the `watch` shim script above, you can see the `PACKAGE_MAIN_OPTS_*` variables being filled in. The `--main-opts` option allows the user to override these variables. If there is no `--main-opts` provided, `bbin` will look for the main opts in `bb.edn` at the `:bin` path (probably going to change the key, as mentioned above), or it will finally fall back to inferring the main opts. For example, if the lib is `io.github.my-ns/my-name`, the main opts will be inferred as `["-m" "my-ns.my-name"]`.

I added the `--main-opts` override to be able to support `neil` without any changes to the existing repo. [If we add a `:bin` key to the `bb.edn` of `neil`](https://github.com/rads/neil/blob/v0.1.44-alpha7/bb.edn#L4), the `--main-opts` override is no longer needed since we can put the `["-f" "neil"]` in there:

```
$ bbin install io.github.rads/neil --git/tag v0.1.44-alpha7 --as neil-alpha
{:coords {:git/sha "96fb18d2e6ac40c3f583814b4c56486da7946c8b",
          :git/tag "v0.1.44-alpha7",
          :git/url "https://github.com/rads/neil"},
 :lib io.github.rads/neil}

$ neil-alpha version
{:neil "v0.1.44-alpha7", :project "0.0.1"}
```
[**Gist:** See the output `neil-alpha` script from the command above](https://gist.github.com/rads/8fa232f51dd113e8ab9ed3a48ab32aee)

Note that this currently only applies to `tools.deps`-based dependencies as opposed to scripts over HTTP (which are also supported as `:http/url` coords, bypassing `tools.deps`). Also, the `bb.edn` config only works with [Gitlibs](https://github.com/clojure/tools.gitlibs) right now.

**Reply by rads** [reply]
Also, [here is an example of the `portal` script from an `:http/url`-based coordinate](https://gist.github.com/rads/d7678731a15146f7b1ccbafb357bfd90). The only difference from [the original Gist](https://gist.github.com/rads/da8ecbce63fe305f3520637810ff9506) is that `bbin` adds some metadata as a comment for `bbin ls` and potentially other commands in the future.

**Reply by borkdude** [reply]
From Slack:

> @borkdude: I think a good way to get a feel for what bb install should look like is to build a prototype in user space like @teodorlu did
> And then we can take the best of all approaches as a built-in

It's good to see how this pans out in user space and then taking the best approach as a built-in, so we don't have to make breaking changes anymore. So thanks for doing this.

If this (or a similar) approach will end up in bb, Windows also has to be supported, but that is doable.

About the [thread](https://github.com/babashka/babashka/discussions/1307#discussioncomment-3388780): I think you may have misunderstood me when I said it's not a general replacement for npm or brew: those allows you to install arbitrary binaries, like glibc, or rustc, or whatever: that was a thing that's not in scope for bb install since this adds extra complexity: bb install focuses on installing bb projects as scripts. So what you describe above is still in scope, AFAIC.

**Reply by rads** [reply]
@borkdude:
> If this (or a similar) approach will end up in bb, Windows also has to be supported, but that is doable

Yep, that is on my list. I don't know much about Windows shell scripting right now but I should be able to figure it out since the Bash version is pretty minimal.

> So what you describe above is still in scope, AFAIC.

Cool! Thanks for the clarification on the earlier discussion. I agree about not supporting installing arbitrary stuff.

I'm going to keep working to get the `bbin@0.0.1` release out so we have a more solid foundation for further discussion.

**Reply by teodorlu** [reply]
For now, I just want to say that this is super exciting :smile: 

> The key thing here is that I don't have to invest anything as a project maintainer to take advantage of this, yet I also don't need to go out of my way as a project consumer to accommodate this. For me personally, this will allow me to release smaller, more focused libraries since it removes the burden of packaging all these things separately.

Yes. If we can make it "free" to share a bb script, that's absolutely amazing. Babashka fixes startup time and dependency resolution. If we can avoid extra mental overhead on top of that, that's just great.

> I've been using this tool myself to install all the Babashka scripts and projects I use.

I also really second trying this out for a while before making public API guarantees! That really makes the extension of core bb functionality less risky.

**Reply by rads** [reply]
@borkdude @teodorlu: [The first release of `bbin` is now available.](https://github.com/rads/bbin) While `v0.0.1` is still very much a work-in-progress, I think it's also pretty useful. Disclaimer: The implementation is messy and there's still a decent amount of work to do. That said, I set up GitHub Actions and wrote some initial tests so I can add things incrementally.

There are a lot of cases to think about with all the combinations of [procurers](https://clojure.org/reference/deps_and_cli#_glossary) (Git, HTTP, Maven, Local) and dependency types (file, directory, JAR), so not every Babashka project is going to work with `v0.0.1`. Eventually I think all the `TODO` examples in the Usage section of the README can be supported.

I also want to be [conscious of security](https://github.com/rads/bbin/blob/main/docs/design.md). If not done carefully, `bbin install` could potentially become an attack vector.

For `bbin`-specific feedback as it relates to `bb install`, I created a new thread: https://github.com/rads/bbin/discussions/6. You're still welcome to discuss things here in the babashka repo. I just wanted to provide an alternative place so we don't clog up this discussion with only `bbin`-specific stuff.

**Reply by borkdude** [reply]
Very cool! I'm on vacation right now so I have limited time available, but I'm very excited to try it out soon.

**Reply by rads** [reply]
All good, no rush 🙂

**Reply by didibus** [reply]
This is pretty cool, it's what I wanted Clojure CLI to do as well (create a shim wrapper around tools so I don't have to prefix them) and also automatically add itself.
